unitsyncer/sansio_lsp_syncer.py

```diff
@@ -0,0 +1,332 @@
+                raise self.exception
+
+            for msg in self.msgs:
+                if isinstance(msg, type_):
+                    self.msgs.remove(msg)
+                    return msg
+
+            time.sleep(0.2)
+
+        raise Exception(
+            f"Didn't receive {type_} in time; have: " + pprint.pformat(self.msgs)
+        )
+
+    def exit_cleanly(self):
+        # Not necessarily error, gopls sends logging messages for example
+        #        if self.msgs:
+        #            print(
+        #                "* unprocessed messages: " + pprint.pformat(self.msgs)
+        #            )
+
+        assert self.lsp_client.state == lsp.ClientState.NORMAL
+        self.lsp_client.shutdown()
+        self.wait_for_message_of_type(lsp.Shutdown)
+        self.lsp_client.exit()
+        self._queue_data_to_send()
+        self._read_data_received()
+
+    def do_method(self, text, file_uri, method, pos, response_type=None):
+        def doc_pos():
+            return lsp.TextDocumentPosition(
+                textDocument=lsp.TextDocumentIdentifier(uri=file_uri),
+                position=pos,
+            )
+
+        if not response_type:
+            response_type = RESPONSE_TYPES[method]
+
+        if method == METHOD_COMPLETION:
+            event_id = self.lsp_client.completion(
+                text_document_position=doc_pos(),
+                context=lsp.CompletionContext(
+                    triggerKind=lsp.CompletionTriggerKind.INVOKED, triggerCharacter=None
+                ),
+            )
+        elif method == METHOD_HOVER:
+            event_id = self.lsp_client.hover(text_document_position=doc_pos())
+        elif method == METHOD_SIG_HELP:
+            event_id = self.lsp_client.signatureHelp(text_document_position=doc_pos())
+        elif method == METHOD_DEFINITION:
+            event_id = self.lsp_client.definition(text_document_position=doc_pos())
+        elif method == METHOD_REFERENCES:
+            event_id = self.lsp_client.references(text_document_position=doc_pos())
+        elif method == METHOD_IMPLEMENTATION:
+            event_id = self.lsp_client.implementation(text_document_position=doc_pos())
+        elif method == METHOD_DECLARATION:
+            event_id = self.lsp_client.declaration(text_document_position=doc_pos())
+        elif method == METHOD_TYPEDEF:
+            event_id = self.lsp_client.typeDefinition(text_document_position=doc_pos())
+        elif method == METHOD_DOC_SYMBOLS:
+            _docid = lsp.TextDocumentIdentifier(uri=file_uri)
+            event_id = self.lsp_client.documentSymbol(text_document=_docid)
+        else:
+            raise NotImplementedError(method)
+
+        resp = self.wait_for_message_of_type(response_type)
+        assert not hasattr(resp, "message_id") or resp.message_id == event_id
+        return resp
+
+
+# ============================ ends utils ========================
+
+
+class SansioLSPSynchronizer(Synchronizer):
```

Comment:
@DataCorrupted should we try if using this client solves the rust LSP error? `rust-analyzer` was showing similar error messages as Go.

Reply:
Why not?